Modules/_pickle.c

```diff
@@ -3965,40 +3962,37 @@ save(PicklerObject *self, PyObject *obj, int pers_save)
     /* Atom types; these aren't memoized, so don't check the memo. */
 
     if (obj == Py_None) {
-        status = save_none(self, obj);
-        goto done;
+        return save_none(self, obj);
     }
     else if (obj == Py_False || obj == Py_True) {
-        status = save_bool(self, obj);
-        goto done;
+        return save_bool(self, obj);
     }
     else if (type == &PyLong_Type) {
-        status = save_long(self, obj);
-        goto done;
+        return save_long(self, obj);
     }
     else if (type == &PyFloat_Type) {
-        status = save_float(self, obj);
-        goto done;
+        return save_float(self, obj);
     }
 
     /* Check the memo to see if it has the object. If so, generate
        a GET (or BINGET) opcode, instead of pickling the object
        once again. */
     if (PyMemoTable_Get(self->memo, obj)) {
-        if (memo_get(self, obj) < 0)
-            goto error;
-        goto done;
+        return memo_get(self, obj);
     }
 
     if (type == &PyBytes_Type) {
-        status = save_bytes(self, obj);
-        goto done;
+        return save_bytes(self, obj);
     }
     else if (type == &PyUnicode_Type) {
-        status = save_unicode(self, obj);
-        goto done;
+        return save_unicode(self, obj);
     }
-    else if (type == &PyDict_Type) {
+
+    if (Py_EnterRecursiveCall(" while pickling an object")) {
```

Comment:
Perhaps add a comment such as """We're only calling Py_EnterRecursiveCall here so that atomic types above are pickled faster"""?

Reply:
Thanks!